<?xml version="1.0"?>
<sdf version="1.5">
<model name="velodyne_hdl-32">
  <!-- Give the base link a unique name -->
  <link name="base">

    <!-- Offset the base by half the lenght of the cylinder -->
    <pose>0 0 0.029335 0 0 0</pose>
    <inertial>
      <mass>1.2</mass>
      <inertia>
        <ixx>0.001087473</ixx>
        <iyy>0.001087473</iyy>
        <izz>0.001092437</izz>
        <ixy>0</ixy>
        <ixz>0</ixz>
        <iyz>0</iyz>
      </inertia>
    </inertial>
    <collision name="base_collision">
      <geometry>
        <cylinder>
          <!-- Radius and length provided by Velodyne -->
          <radius>.04267</radius>
          <length>.05867</length>
        </cylinder>
      </geometry>
    </collision>

    <!-- The visual is mostly a copy of the collision -->
    <visual name="base_visual">
      <!-- Offset the visual by have the base's height. We are not rotating
           mesh since symmetrical -->
      <pose>0 0 -0.029335 0 0 0</pose>
      <geometry>
        <mesh>
          <uri>model://velodyne_hdl32/meshes/velodyne_base.dae</uri>
        </mesh>
      </geometry>
    </visual>
  </link>

    <link name="top">
        <sensor type="ray" name="sensor">
            <pose>0 0 -0.004645 0 0 0</pose>
            <visualize>true</visualize>
            <update_rate>10</update_rate>
            <ray>
                <noise>
                    <type>gaussian</type>
                    <mean>0.0</mean>
                    <stddev>0.02</stddev>
                </noise>
                <scan>
                    <horizontal>
                        <samples>2187</samples>
                        <resolution>1</resolution>
                        <min_angle>-3.1415926</min_angle>
                        <max_angle>3.1415926</max_angle>
                    </horizontal>

                    <vertical>
                        <samples>32</samples>
                        <resolution>1</resolution>
                        <min_angle>-0.53529248</min_angle>
                        <max_angle>0.18622663</max_angle>
                    </vertical>
                </scan>
                <range>
                    <min>0.05</min>
                    <max>70</max>
                    <resolution>0.02</resolution>
                </range>
            </ray>
        </sensor>

        <pose>0 0 0.095455 0 0 0</pose>
        <inertial>
            <mass>0.1</mass>
                <inertia>
                    <ixx>0.000090623</ixx>
                    <iyy>0.000090623</iyy>
                    <izz>0.000091036</izz>
                    <ixy>0</ixy>
                    <ixz>0</ixz>
                    <iyz>0</iyz>
                </inertia>
        </inertial>
    <collision name="top_collision">
      <geometry>
        <cylinder>
          <!-- Radius and length provided by Velodyne -->
          <radius>0.04267</radius>
          <length>0.07357</length>
        </cylinder>
      </geometry>
    </collision>

    <!-- The visual is mostly a copy of the collision -->
    <visual name="top_visual">
      <!-- Lower the mesh by half the height, and rotate by 90 degrees -->
      <pose>0 0 -0.0376785 0 0 1.5707</pose>
      <geometry>
        <!-- The mesh tag indicates that we will use a 3D mesh as
             a visual -->
        <mesh>
          <!-- The URI should refer to the 3D mesh. The "model:"
              URI scheme indicates that the we are referencing a Gazebo
              model. -->
          <uri>model://velodyne_hdl32/meshes/velodyne_top.dae</uri>
        </mesh>
      </geometry>
    </visual>
  </link>

    <joint type="fixed" name="joint">
        <pose>0 0 -0.036785 0 0 0</pose>
        <parent>base</parent>
        <child>top</child>
    </joint>
</model>
</sdf>
